nalgebra-sparse/src/io/matrix_market.rs

```diff
@@ -0,0 +1,861 @@
+        .next()
+        .unwrap()
+        .as_str()
+        .to_ascii_lowercase()
+        .parse::<DataType>()
+        .unwrap();
+    header_type.ss = inner
+        .next()
+        .unwrap()
+        .as_str()
+        .to_ascii_lowercase()
+        .parse::<StorageScheme>()
+        .unwrap();
+    typecode_precheck(&header_type)
+}
+
+/// Parsing a pest structure to 3 int, which are number of rols, cols and non-zeros.
+fn parsing_shape(inner: &mut Pairs<'_, Rule>, shape: &mut (usize, usize, usize)) {
+    // once the file can be parsed, all unwrap() in this function here will not panic.
+    shape.0 = inner.next().unwrap().as_str().parse::<usize>().unwrap();
+    shape.1 = inner.next().unwrap().as_str().parse::<usize>().unwrap();
+    shape.2 = inner.next().unwrap().as_str().parse::<usize>().unwrap();
+}
+
+#[inline]
+/// Do a precheck of remaing data. Because in matrix_market.pest file, either 0, or 1 or 2 float number(s) as data is valid, however, for example, for Real matrix, only 1 float is valid. And 2 float for Complex matrix.  
+fn check_value_length(inner: &Pairs<'_, Rule>, l: usize) -> Result<(), MMError> {
+    let copy = inner.clone();
+    let c = copy.count();
+    if l != c {
+        return Err(MMError::from_kind_and_message(
+            MMErrorKind::ParsingError,
+            format!("{} data required, but {} provided.", l, c),
+        ));
+    }
+    Ok(())
+}
+
+#[inline]
+// do a quick check it the entry is in the lower triangle part of the matrix
+fn check_lower_triangle(r: usize, c: usize) -> Result<(), MMError> {
+    if c > r {
+        return Err(MMError::from_kind_and_message(
+            MMErrorKind::LowerTriangleError,
+            format!(
+                "Entry: row {} col {} should be put into lower triangle",
+                r, c
+            ),
+        ));
+    }
+    Ok(())
+}
+
+/// Parses a Matrix Market file at the given path, and returns the corresponding sparse matrix as CooMatrix format.
+pub fn load_coo_from_mm_file<T, P: AsRef<Path>>(path: P) -> Result<CooMatrix<T>, MMError>
+where
+    T: MMType,
+{
+    let file = fs::read_to_string(path)?;
+    load_coo_from_mm_str(&file)
+}
+
+/// Parses a Matrix Market file described by the given string, and returns the corresponding as CooMatrix format.
+pub fn load_coo_from_mm_str<T>(data: &str) -> Result<CooMatrix<T>, MMError>
+where
+    T: MMType,
+{
+    // unwrap() here guaranteed when data can be parsed
+    let file = MMParser::parse(Rule::Document, data)?.next().unwrap();
+    //  Default typecode
+    let mut header_type = Typecode {
+        sp: Sparsity::Sparse,
+        dt: DataType::Real,
+        ss: StorageScheme::General,
+    };
+    let mut shape = (0, 0, 0);
+    let mut rows: Vec<usize> = Vec::new();
+    let mut cols: Vec<usize> = Vec::new();
+    let mut data: Vec<T> = Vec::new();
+
+    let mut count = 0;
+    for line in file.into_inner() {
+        match line.as_rule() {
+            Rule::Header => {
```

Comment:
To me it's a little bit strange that we look for a header for each line in the whole file: wouldn't it be more natural to first look for header & shape, then parse only entries after that?